docs/JAX_basic_text_classification.md

```diff
@@ -0,0 +1,455 @@
+```
+
+```{code-cell} ipython3
+def compute_losses_and_logits(model: nnx.Module, batch_tokens: jax.Array, labels: jax.Array):
+    logits = model(batch_tokens)
+
+    loss = optax.softmax_cross_entropy_with_integer_labels(
+        logits=logits, labels=labels
+    ).mean()
+    return loss, logits
+```
+
+```{code-cell} ipython3
+@nnx.jit
+def train_step(
+    model: nnx.Module, optimizer: nnx.Optimizer, batch: dict[str, jax.Array]
+):
+    # Convert numpy arrays to jax.Array on GPU
+    batch_tokens = jnp.array(batch["text"])
+    labels = jnp.array(batch["label"], dtype=jnp.int32)
+
+    grad_fn = nnx.value_and_grad(compute_losses_and_logits, has_aux=True)
+    (loss, logits), grads = grad_fn(model, batch_tokens, labels)
+
+    optimizer.update(grads)  # In-place updates.
+
+    return loss
+
+
+@nnx.jit
+def eval_step(
+    model: nnx.Module, batch: dict[str, jax.Array], eval_metrics: nnx.MultiMetric
+):
+    # Convert numpy arrays to jax.Array on GPU
+    batch_tokens = jnp.array(batch["text"])
+    labels = jnp.array(batch["label"], dtype=jnp.int32)
+    loss, logits = compute_losses_and_logits(model, batch_tokens, labels)
+
+    eval_metrics.update(
+        loss=loss,
+        logits=logits,
+        labels=labels,
+    )
+```
+
+```{code-cell} ipython3
+eval_metrics = nnx.MultiMetric(
+    loss=nnx.metrics.Average('loss'),
+    accuracy=nnx.metrics.Accuracy(),
+)
+
+
+train_metrics_history = {
+    "train_loss": [],
+}
+
+eval_metrics_history = {
+    "test_loss": [],
+    "test_accuracy": [],
+}
+```
+
+```{code-cell} ipython3
+import tqdm
+
+
+bar_format = "{desc}[{n_fmt}/{total_fmt}]{postfix} [{elapsed}<{remaining}]"
+train_total_steps = len(train_dataset) // train_batch_size
+
+
+def train_one_epoch(epoch):
+    model.train()  # Set model to the training mode: e.g. update batch statistics
+    with tqdm.tqdm(
+        desc=f"[train] epoch: {epoch}/{num_epochs}, ",
+        total=train_total_steps,
+        bar_format=bar_format,
+        leave=True,
+    ) as pbar:
+        for batch in train_loader:
+            loss = train_step(model, optimizer, batch)
+            train_metrics_history["train_loss"].append(loss.item())
+            pbar.set_postfix({"loss": loss.item()})
+            pbar.update(1)
+
+
+def evaluate_model(epoch):
+    # Compute the metrics on the train and val sets after each training epoch.
+    model.eval()  # Set model to evaluation model: e.g. use stored batch statistics
+
+    eval_metrics.reset()  # Reset the eval metrics
+    for test_batch in test_loader:
+        eval_step(model, test_batch, eval_metrics)
+
+    for metric, value in eval_metrics.compute().items():
+        eval_metrics_history[f'test_{metric}'].append(value)
+
+    print(f"[test] epoch: {epoch + 1}/{num_epochs}")
+    print(f"- total loss: {eval_metrics_history['test_loss'][-1]:0.4f}")
+    print(f"- Accuracy: {eval_metrics_history['test_accuracy'][-1]:0.4f}")
+```
+
+Let us now start the training
```

Comment:
Just a suggestion:
```suggestion
Now, we can start the training. Note that this step can take quite a while to run, depending on the number of epochs you are using. 
```

Reply:
Well, this is not exactly true, in this case, whole training takes 2 minutes. The model and the dataset are rather small.

Reply:
Interesting! For me it took something like 40 mins in colab.